language-plutus-core/exe/Main.hs

```diff
@@ -54,16 +62,55 @@ stdInput = flag' StdInput
   (  long "stdin"
   <> help "Read from stdin" )
 
+
+data Output = FileOutput FilePath | StdOutput
+
+output :: Parser Output
+output = fileOutput <|> stdOutput
+
+fileOutput :: Parser Output
+fileOutput = FileOutput <$> strOption
+  (  long "output"
+  <> short 'o'
+  <> metavar "FILENAME"
+  <> help "Output file" )
+
+stdOutput :: Parser Output
+stdOutput = flag' StdOutput
+  (  long "stdout"
+  <> help "Write to stdout" )
+
+
+data Format = Plc | Cbor  -- Input/output format for programs
+
+format :: Parser Format
+format = flag Plc Cbor
+  ( long "cbor"
+  <> long "CBOR"
+  <> short 'c'
+  <> help "Input CBOR (default: input PLC)"
+  )
+
+
 data NormalizationMode = Required | NotRequired deriving (Show, Read)
-newtype TypecheckOptions = TypecheckOptions Input
+data TypecheckOptions = TypecheckOptions Input Format
 data EvalMode = CK | CEK deriving (Show, Read)
-data EvalOptions = EvalOptions Input EvalMode
+data EvalOptions = EvalOptions Input EvalMode Format
+data PrintMode = Classic | Debug deriving (Show, Read)
```

Comment:
Why isn't CBOR simply a `PrintMode`?

Reply:
Summing up, why not:
- `data Format = Classic | Readable | CBOR`
- `data PrintMode = Normal | Debug`

Then have an `inputFormat` and an `outputFormat` from the same set. We don't have a parser for `Readable` so you just error in that case.

Reply:
I don't quite get this.  We could have a single `Format` type and use it in different places (not using all of the cases everywhere), but we'd still need the same number of parsers.

Suppose you have `Format` as above and you want a single parser for it, which can return any of the values.  You'd need an option name like `--format`, and then if you want CBOR you'd have to type `--format cbor`.  I wouldn't want to do this if I want to evaluate a CBOR program: I just want to type `--cbor`, and that requires a separate parser.  Furthermore, if we _did_ use the three-option parser everywhere, `--help` would advertise all three options even if two of them were going to do exactly the same thing (if we're reading PLC we don't care about the format), and that would be confusing.  

So I think we'd still need exactly the same number of parsers, except that each one would return a value of type `Format`, and in some cases certain return values would be impossible (and we'd still have to add code to deal with impossible cases).  I'm not convinced that that's an improvement.

Reply:
I'm suggesting we have `--input-format` and `--output-format`, both of which accept the same enum of options. That seems more regular to me. And we do clear stuff up: we don't need the new commands, because you just do `plc print --input-format CBOR --output-format classic` or vice versa.

Yes, `--input-format CBOR` is longer, but  Iike the regularity. I wouldn't hate it if we had `--cbor` as a shorthand. But I do think it's better to get rid of the excess commands.

Reply:
I did consider that, but it seemed a bit too verbose.  You'd have to type things like
```
  plc convert --input factorial.plc --input-format Classic --output factorial.cbor --output-format CBOR
```
That's quite a lot of typing, and it just seems like subjecting the user to inconvenience for the sake of logical purity.  I think that having a separate `plc-to-cbor` command does what the used would expect in a concise way, and there's plenty of help to explain the details.  Also, allowing eg CBOR-to-CBOR conversions seems a bit odd.

Reply:
Well, surely it's
```
plc print -i factorial.plc -o factorial.cbor --output-format CBOR
``
which doesn't seem so bad (classic is the default).

If we want to be fancy we could autodetect the input/output formats from the file extension if they're not specified.

> and it just seems like subjecting the user to inconvenience for the sake of logical purity.

I find it much easier to remember a set of consistent flags than a variety of specialized subcommands, personally. 

> Also, allowing eg CBOR-to-CBOR conversions seems a bit odd.

That's why I'm just suggesting it's just a feature of `print`. Printing a CBOR file as CBOR indeed does nothing, but that's what you asked for. I don't think that's particularly odd.